### PR TITLE
add fuzzer dir for each apigroup

### PR DESCRIPTION
**Pull request by zhouhaibing089**
Fixes https://github.com/kubernetes/kubernetes/issues/50081.

We should put fuzzer functions in each apigroups separately to make the code more modular instead of monolithic.

### CODE REVIEW

**Comment by sttts** [comment]
Lgtm. Waiting for @caesarxuchao's opinion though.

File: pkg/apis/abac/fuzzer/fuzzer.go [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+	"k8s.io/apimachinery/pkg/api/testing/fuzzer"
+	runtimeserializer "k8s.io/apimachinery/pkg/runtime/serializer"
+)
+
+func abacFuncs(codecs runtimeserializer.CodecFactory) []interface{} {
+	return []interface{}{}
+}
+
+// Funcs returns the fuzzer functions for the abac api group.
+var Funcs = fuzzer.MergeFuzzerFuncs(
```
Comment (ncdc) [inline pkg/apis/abac/fuzzer/fuzzer.go]: Is there a need to call merge here - how about just renaming `abacFuncs` to `Funcs`? Same question for all the other new files as well.

Comment (sttts) [inline pkg/apis/abac/fuzzer/fuzzer.go]: I agree, we should get rid of all these `fuzzer.MergeFuzzerFuncs` calls. I think @zhouhaibing089 is already on it.

File: pkg/apis/admission/install/roundtrip_test.go [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+import (
+	"testing"
+
+	"k8s.io/apimachinery/pkg/api/testing/roundtrip"
+	admissionfuzzer "k8s.io/kubernetes/pkg/apis/admission/fuzzer"
+)
+
+func TestRoundTrip(t *testing.T) {
+	roundtrip.RoundTripTestForAPIGroup(t, Install, admissionfuzzer.Funcs)
```
Comment (caesarxuchao) [inline pkg/apis/admission/install/roundtrip_test.go]: Why do we need a `roundtrip_test.go` in each group? Isn't the centralized [roundtrip](https://github.com/kubernetes/kubernetes/blob/master/pkg/api/serialization_test.go#L206) test enough? We shouldn't distribute the burden of writing roundtrip tests to each API developer.

Breaking the monolithic fuzzer file makes sense.

Comment (zhouhaibing089) [inline pkg/apis/admission/install/roundtrip_test.go]: @caesarxuchao well, the centralized roundtrip test is enough to capture all the api types imported from `k8s.io/kubernetes/pkg/api/testap`.

I think the problem is whether we are consistent everywhere, see:

```console
$ find staging/* -name roundtrip_test.go
staging/src/k8s.io/apiextensions-apiserver/examples/client-go/apis/cr/v1/roundtrip_test.go
staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/install/roundtrip_test.go
staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/roundtrip_test.go
staging/src/k8s.io/apimachinery/pkg/apis/testapigroup/install/roundtrip_test.go
staging/src/k8s.io/apiserver/pkg/apis/example/install/roundtrip_test.go
staging/src/k8s.io/sample-apiserver/pkg/apis/wardle/install/roundtrip_test.go
```

Comment (caesarxuchao) [inline pkg/apis/admission/install/roundtrip_test.go]: I understand why the staging repos need the `roundtrip_test.go`. It's unnecessary to have `roundtrip_test.go` in each kubernetes API group.

Comment (zhouhaibing089) [inline pkg/apis/admission/install/roundtrip_test.go]: got it, I'll remove the test then..

**Comment by zhouhaibing089** [comment]
@ncdc @caesarxuchao @sttts removed the `roundtrip_test.go` under each api group and updated to use `Funcs` name.

File: federation/registry/cluster/strategy_test.go [hunk 3]
```diff
@@ -27,6 +27,7 @@ import (
 	genericapirequest "k8s.io/apiserver/pkg/endpoints/request"
 	"k8s.io/kubernetes/federation/apis/federation"
 	"k8s.io/kubernetes/pkg/api"
+	_ "k8s.io/kubernetes/pkg/api/testapi"
```
Comment (zhouhaibing089) [inline federation/registry/cluster/strategy_test.go]: I moved the `coreFuncs` to `pkg/api/fuzzer`, and thus `pkg/api/testing` no longer imports `pkg/api/testapi`(which registers all the kubernetes groups).

and I think it is better to make it explicit that these `strategy_test.go` requires `pkg/api/testapi`, so I add it here with an underscore import.

Comment (sttts) [inline federation/registry/cluster/strategy_test.go]: Not sure I understand. Why is `pkg/api/testing` needed here?

File: pkg/apis/abac/fuzzer/fuzzer.go [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+*/
+
+package fuzzer
+
+import (
+	runtimeserializer "k8s.io/apimachinery/pkg/runtime/serializer"
+)
+
+// Funcs returns the fuzzer functions for the abac api group.
+var Funcs = func(codecs runtimeserializer.CodecFactory) []interface{} {
```
Comment (sttts) [inline pkg/apis/abac/fuzzer/fuzzer.go]: much better now with the merge :+1:

File: pkg/api/testing/fuzzer.go [hunk 5]
```diff
@@ -94,699 +87,17 @@ func overrideGenericFuncs(codecs runtimeserializer.CodecFactory) []interface{} {
...
 var FuzzerFuncs = fuzzer.MergeFuzzerFuncs(
 	genericfuzzer.Funcs,
 	overrideGenericFuncs,
-	coreFuncs,
-	extensionFuncs,
-	appsFuncs,
-	batchFuncs,
-	autoscalingFuncs,
-	rbacFuncs,
+	corefuzzer.Funcs,
```
Comment (sttts) [inline pkg/api/testing/fuzzer.go]: nice and short :+1:

File: federation/registry/cluster/strategy_test.go (hunk 3 above)
Comment (zhouhaibing089) [inline federation/registry/cluster/strategy_test.go]: See an example [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/core/node/strategy_test.go#L50), the call to `api.Registry.GroupOrDie(api.GroupName)` would die if no group registered(it won't without importing `pkg/api/testapi`).

Comment (sttts) [inline federation/registry/cluster/strategy_test.go]: But there is line below? Why the anonymous import?

Comment (zhouhaibing089) [inline federation/registry/cluster/strategy_test.go]: Along with this move: `pkg/api/testing/fuzzer.go` to `pkg/api/fuzzer/fuzzer.go`, the import is also moved to `pkg/api/fuzzer/fuzzer.go`(remember the EnvVar randomization?).

as a result, importing `pkg/api/testing` does not imply importing `pkg/api/testapi` any longer.

Comment (sttts) [inline federation/registry/cluster/strategy_test.go]: Oh, `testing` and `testapi` read nearly the same. Makes sense then.

Can you separate that import line somehow like this:
```
...
apitesting "k8s.io/kubernetes/pkg/api/testing"

// install all api groups for testing
_ "k8s.io/kubernetes/pkg/api/testapi"
```

Comment (zhouhaibing089) [inline federation/registry/cluster/strategy_test.go]: done.

**Comment by sttts** [comment]
/assign @smarterclayton 

Another one for your to approve.

File: federation/registry/cluster/strategy_test.go [hunk 6]
```diff
@@ -28,6 +28,9 @@ import (
 	"k8s.io/kubernetes/federation/apis/federation"
 	"k8s.io/kubernetes/pkg/api"
 	apitesting "k8s.io/kubernetes/pkg/api/testing"
+
+	// install all api groups for testing
+	_ "k8s.io/kubernetes/pkg/api/testapi"
```
Comment (caesarxuchao) [inline federation/registry/cluster/strategy_test.go]: Why do you need this change?

File: pkg/api/testing/fuzzer.go [hunk 7]
```diff
@@ -18,35 +18,28 @@ package testing
...
-	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	"k8s.io/apimachinery/pkg/fields"
-	"k8s.io/apimachinery/pkg/labels"
 	"k8s.io/apimachinery/pkg/runtime"
-	"k8s.io/apimachinery/pkg/runtime/schema"
 	runtimeserializer "k8s.io/apimachinery/pkg/runtime/serializer"
-	"k8s.io/apimachinery/pkg/util/intstr"
 	kubeadmfuzzer "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/fuzzer"
 	"k8s.io/kubernetes/pkg/api"
-	"k8s.io/kubernetes/pkg/api/testapi"
```
Comment (caesarxuchao) [inline pkg/api/testing/fuzzer.go]: Probably we were relying on the side effect of importing `k8s.io/kubernetes/pkg/api/testapi`. Try adding it back and see if the unit tests passed.

File: federation/registry/cluster/strategy_test.go (hunk 3 above)
Comment (caesarxuchao) [inline federation/registry/cluster/strategy_test.go]: @sttts do you prefer importing `"k8s.io/kubernetes/pkg/api/testing"` in eatch `_test.go` or just importing it in `pkg/api/testing`? I can see arguments both sides.

File: federation/registry/cluster/strategy_test.go (hunk 6 above)
Comment (caesarxuchao) [inline federation/registry/cluster/strategy_test.go]: Please ignore, a github UI bug.

**Comment by caesarxuchao** [comment]
@zhouhaibing089 @sttts actually I do prefer centralizing the import of `"k8s.io/kubernetes/pkg/api/testapi"` in `pkg/api/testing`. Could you revert the changes? I'm taking off the lgtm for now.

**Comment by zhouhaibing089** [comment]
@caesarxuchao actually, I have different opinion for the following reasons:

1. the reason that `pkg/api/testing` imports `pkg/api/testapi` before is that [EnvVar randomization](https://github.com/kubernetes/kubernetes/blob/master/pkg/api/testing/fuzzer.go#L355) relies on the group versions, so I did not see the side effect is intentional, it looks like more accidental.
2. I also tried adding the `_` import in `pkg/api/testing`, however my editor complains(though just warnings) that `_` import should happen only in main package and test files.
3. more explicitly on the imports, more easier for the future changes.

Does it make sense?

**Comment by caesarxuchao** [comment]
Let's revisit when we refactor the testapi. I'm adding back the lgtm.

**Comment by zhouhaibing089** [comment]
Rebased according to https://github.com/kubernetes/kubernetes/pull/49983

**Pull request merged**